src/kvstore/comm.h

```diff
@@ -178,85 +179,94 @@ class CommCPU : public Comm {
         const_vars[i] = reduce[i].var();
       }
       NDArray result = buf.merged;
-      Resource rsc = ResourceManager::Get()->Request(result.ctx(),
-          ResourceRequest(ResourceRequest::kTempSpace));
+      Resource rsc = ResourceManager::Get()->Request(
+          result.ctx(), ResourceRequest(ResourceRequest::kTempSpace));
       Engine::Get()->PushAsync(
-        [reduce, result, rsc, this](RunContext rctx, Engine::CallbackOnComplete on_complete) {
-          NDArray out = result;
-          is_serial_push_?
-            ReduceSumCPUExSerial(reduce, &out)
-            : mxnet::ndarray::ElementwiseSum(rctx.get_stream<cpu>(), rsc, reduce, &out);
-          on_complete();
-        }, Context::CPU(), const_vars, {result.var(), rsc.var},
-        FnProperty::kCPUPrioritized, priority, PROFILER_MESSAGE("KVStoreReduce"));
+          [reduce, result, rsc, this](RunContext rctx,
+                                      Engine::CallbackOnComplete on_complete) {
+            NDArray out = result;
+            is_serial_push_ ? ReduceSumCPUExSerial(reduce, &out)
+                            : mxnet::ndarray::ElementwiseSum(
+                                  rctx.get_stream<cpu>(), rsc, reduce, &out);
+            on_complete();
+          },
+          Context::CPU(), const_vars, {result.var(), rsc.var},
+          FnProperty::kCPUPrioritized, priority,
+          PROFILER_MESSAGE("KVStoreReduce"));
     }
 
     return buf.merged;
   }
 
-  void Broadcast(int key, const NDArray& src,
-                 const std::vector<NDArray*> dst, int priority) override {
+  void Broadcast(int key, const NDArray& src, const std::vector<NDArray*> dst,
+                 int priority) override {
     int mask = src.ctx().dev_mask();
     if (mask == Context::kCPU) {
-      for (auto d : dst) CopyFromTo(src, d, priority);
+      for (auto& d : dst) CopyFromTo(src, d, priority);
```

Comment:
Is `broadcast` not leveraging the new comm pattern? The same logic can be applied to fully utilize the bandwidth during copy, right? Or is the plan to do that in the next PR?